src/library/scala/runtime/BoxesRunTime.java

```diff
@@ -117,36 +117,29 @@ public static double unboxToDouble(Object d) {
     /* COMPARISON ... COMPARISON ... COMPARISON ... COMPARISON ... COMPARISON ... COMPARISON */
 
     public static boolean equals(Object x, Object y) {
-        if (x == y) return true;
-        return equals2(x, y);
+        return (x == y) || equalsNotSame(x, y);
     }
 
-    /** Since all applicable logic has to be present in the equals method of a ScalaNumber
-     *  in any case, we dispatch to it as soon as we spot one on either side.
-     */
-    public static boolean equals2(Object x, Object y) {
-        if (x instanceof java.lang.Number)
-            return equalsNumObject((java.lang.Number)x, y);
-        if (x instanceof java.lang.Character)
-            return equalsCharObject((java.lang.Character)x, y);
-        if (x == null)
-            return y == null;
-
-        return x.equals(y);
+    private static boolean equalsNotSame(Object x, Object y) {
+        return x != null && y != null && equalsNotSameOrNull(x, y);
     }
 
-    public static boolean equalsNumObject(java.lang.Number xn, Object y) {
-        if (y instanceof java.lang.Number)
-            return equalsNumNum(xn, (java.lang.Number)y);
-        if (y instanceof java.lang.Character)
-            return equalsNumChar(xn, (java.lang.Character)y);
-        if (xn == null)
-            return y == null;
+    private static boolean equalsNotSameOrNull(Object x, Object y) {
+        if (x instanceof Number) return equalsNumObject((Number) x, y);
+        else if (x instanceof Character) return equalsCharObject((Character) x, y);
+        else return x.equals(y);
+    }
 
-        return xn.equals(y);
+    static boolean equalsNumObject(java.lang.Number xn, Object y) {
```

Comment:
It is safe to introduce new private methods, and to refactor existing methods to use those private methods, so long at the semantics of existing non-private methods is unchanged.

I'd suggest to split out changes to `BoxesRuntime` into a separate commit and present benchmark results. We need to weigh up the performance wins against the risk of introducing regressions or subtle compatibility problems when mixing library from 2.12.1 with code compiled with 2.12.0 and vice versa.